RFC: Configuring popper uniformly across v9 (#19147)
microsoft/fluentui: rfcs/convergence/exposing-popper-options.md

There are also additional options that are not used by any converge component that might be used in the future like:

- `autoSize` - Allows the positioned element to change size based on viewport size
- `flipBoundary` - Determines what bounds to use to flip the positioning

Issues and bugs can be triggered because certain use cases require more precise configuration of popper. The most likely
result would be that a popper configuration is exposed for one component but not others.

## Detailed Design or Proposal

### Equal popper configuration

Each converged component that uses popper should allow an equal level of configuration. This ensures that our positioned
components can behave consistently across the library. Any fixes that require exposing another popper configuration will
be available to **all** components that use popper.

This proposal does not mean that all popper options must be public, but that if one component exposes a popper configuration,
then all components should do so equally.

We can achieve this with an agreement that the current `PositioningProps` would be supported by all v9 components, while

> internal popper options can still belong to `PopperOptions` type in the `react-positioning` package.

Reviewer: FYI: I created #19230, `PopperOptions` should be never part of public APIs.

```tsx
import { PositioningProps } from '@fluentui/react-positioning';

export interface TooltipProps extends PositioningProps {}

export interface MenuProps extends PositioningProps {}

export interface PopoverProps extends PositioningProps {}
```

### `positioning` configuration object

The next step to this proposal is to make sure that all components which use a positioned element should allow users

> to configure popper through a `positioning` configuration object:

Reviewer 2: I'm a little confused, in this RFC is the `positioning` object proposal intended to be an alternative to `extends PositioningProps`, or work together with it?

If it's either/or, then I have a slight preference to the `extends PositioningProps` option, since it makes the API for each control a little simpler:
`<Tooltip position="bottom" align="center" />`
instead of:
`<Tooltip positioning={{ position: 'bottom', align: 'center' }} />`

Author: it was intended to be an alternative that might be clearer for users (?) 

It's not that important and I can remove it for now, what's important is that we should be consistent across the library IMO

Reviewer: There is a caveat with `extends PositioningProps`, it means that component authors will need to pass down all props to `usePopper()`:

```tsx
function Tooltip(props) {
  const {
    align,
    flipBoundary,
    overflowBoundary,
    popperRef,
    position,
    positionFixed,
    offset,
    unstable_disableTether,
    unstable_pinned,
    autoSize
  } = props;

  usePopper({
    align,
    flipBoundary,
    overflowBoundary,
    popperRef,
    position,
    positionFixed,
    offset,
    unstable_disableTether,
    unstable_pinned,
    autoSize
  });

  return <div />;
}
```

And usually, it's hard to track usages and it was so annoying that I created a test for it:

https://github.com/microsoft/fluentui/blob/31f949a20e204a266fc3849b128ea0be083c1dda/packages/fluentui/react-northstar/test/specs/commonTests/implementsPopperProps.tsx#L27-L41

⚠ The problem is that it works only with `Popper` component in v0, we don't have similar test for a hook.

---

Overall, it's one of reasons why I would prefer to use `positioning` as a single prop.

Reviewer 2: That makes sense to me. What do you think about naming it `placement` instead of `positioning`?

Reviewer 2: Also, I'm not 100% sure this is a great idea, but what about allowing a shorthand syntax for specifying position or align as a string:
```jsx
placement="above" // placement={{ position: 'above' }}
placement="top" // placement={{ align: 'top' }}
placement="below-start" // placement={{ position: 'below', align: 'start' }}
```

Reviewer: - I think that we should merge `position` & `align` into a single property, `placement` makes sense
- Shorthand-like syntax may work 👍

Author: @behowell  I really like the shorthand idea 🚀 , will make it easy to distinguish between complex and simple configurations

Author: @behowell  @layershifter  I've added the proposal for the shorthand `placement` prop

```tsx
import { PositioningProps } from '@fluentui/react-positioning';

export interface TooltipProps {
  positioning?: PositioningProps;
}

export interface MenuProps {
  positioning?: PositioningProps;
}

export interface PopoverProps {
  positioning?: PositioningProps;
}
```

Slots that use popper should use a similar pattern if a top level `positioning` option is not clear enough:

```tsx
import { PositioningProps } from '@fluentui/react-positioning';

export interface ComponentWithPositionedSlotProps {
  positionedSlot: React.HTMLAttributes<HTMLElement> & { positioning?: PositioningProps };
}
```

### Pros and Cons

#### Pros

- Positioning capability is consistent across the library
- Fixes to positioned elements through extra configuration can be used across the library
- Configuring positioned elements is consistent across components and clear for users

> - If a better underlying library than `Popper` exists, we can replace it by implementing the `PositioniningProps`interface

Reviewer: ```suggestion
- If a better underlying library than `Popper.js` exists, we can replace it by implementing the `PositioniningProps` interface
```

> - Technically a new `positioning`prop is a breaking API change, but can be done in a non-breaking way

Reviewer: ```suggestion
- Technically a new `positioning` prop is a breaking API change, but can be done in a non-breaking way
```

Reviewer: Can you please clarify how?

Author: Added clarification in the proposal for `placement` prop

Reviewer: ```suggestion
- Technically a new `placement` prop is a breaking API change, but can be done in a non-breaking way
```

#### Cons

> - Some components will not need all complex configuiration

Reviewer: ```suggestion
- Some components will not need all complex configuration
```

- Encourage users to use complicated solutions to their problems
- Adding configuration required only by a single component is expensive

> - Technically a new `positioning`prop is a breaking API change, but can be done in a non-breaking way

Reviewer: ```suggestion
- Technically a new `positioning` prop is a breaking API change, but can be done in a non-breaking way
```

Reviewer: ```suggestion
- Technically a new `placement` prop is a breaking API change, but can be done in a non-breaking way
```

then all components should do so equally.

We can achieve this with an agreement that the current `PositioningProps` would be supported by all v9 components, while
internal popper options can still belong to `PopperOptions` type in the `react-positioning` package.

```tsx
import { PositioningProps } from '@fluentui/react-positioning';

export interface TooltipProps extends PositioningProps {}

export interface MenuProps extends PositioningProps {}

export interface PopoverProps extends PositioningProps {}
```

### Shorhand `placement` configuration

The next step to this proposal is to make sure that all components which use a positioned element should allow users
to configure popper through a `placement` prop that is behaves like a shorthand. The usage examples below will detail
the usage of this shorthand.

```tsx
import { PositioningProps } from '@fluentui/react-positioning';

export interface TooltipProps {
  placement?: PositioningProps;
}

export interface MenuProps {
  placement?: PositioningProps;
}

export interface PopoverProps {
  placement?: PositioningProps;
}
```

Example usage:

```tsx

> // Shorthand for `position` and `align` options

Reviewer 2: This isn't specified in this RFC, but since I mentioned it in my earlier comment, I wanted to amend my suggestion about the shorthand: It's probably best _not_ to have align-only shorthand (like ~~`placement="start"`~~), and just have either position-only like `placement="above"` (implies `align: 'center'`), or position-align like `placement="above-start"`. I.e.:
```ts
type PositioningShorthand =
  | 'above'
  | 'above-start'
  | 'above-end'
  | 'below'
  | 'below-start'
  | 'below-end'
  | 'before'
  | 'before-top'
  | 'before-bottom'
  | 'after'
  | 'after-top'
  | 'after-bottom';
```

If you still wanted to only specify align, you could still do it using the longhand syntax `placement={{ align: 'start' }}`, but that's probably not as common of a use-case.

Author: IMO this is implementation detail how we want to implement the shorthand values, but yes I agree, the values must have a specific order

Author: Added your helpful type enumeration

// Also allows more complex configuration
<PositionedComponent placement={{position: 'above', align: 'start', autoSize: true}}>
```

Slots that use popper should use a similar pattern if a top level `placement` option is not clear enough:

```tsx
import { PositioningProps } from '@fluentui/react-positioning';

export interface ComponentWithPositionedSlotProps {
  positionedSlot: React.HTMLAttributes<HTMLElement> & { placement?: PositioningProps };
}
```

This way we can safely configure `usePopper` internally with a simple object spread that does not require knowing what
props are used by the component props:

```tsx
// Need to be aware of user props and make sure they are added to the hook usage
usePopper({
  align: state.align,
  position: state.position,
  target: state.target,
  coverTarget: state.coverTarget,
  offset: state.offset,
});

// Guaranteed to configure based on user props, and any component specific modifications after
usePopper({
  ...placement,
  // add state specific overrides below
  offset: state.modifiedOffset,
});
```

This change can be made in a non-breaking way by deprecating all existing configuration props outside of the proposed

> `placement` prop and providing support until a specific release in the future (beta/initial release/next major).

Author: If we as a team are ok with that breaking change for beta, sure

> `placement` prop and providing support until a specific release in the future (beta/initial release/next major).

Author: also if we implement this before beta, there will still be alpha versions that have a breaking change, not sure if that's important tho

> `placement` prop and providing support until a specific release in the future (beta/initial release/next major).

Author: @behowell  I've added all the relevant implem details and types in this RFC now, please take a look 🙏

```tsx
import { PositioningProps } from '@fluentui/react-positioning';

export interface TooltipProps extends PositioningProps {}

export interface MenuProps extends PositioningProps {}

export interface PopoverProps extends PositioningProps {}
```

### Shorhand `placement` configuration

The next step to this proposal is to make sure that all components which use a positioned element should allow users
to configure popper through a `placement` prop that is behaves like a shorthand. The usage examples below will detail
the usage of this shorthand.

```tsx
import { PositioningProps } from '@fluentui/react-positioning';

export interface TooltipProps {
  placement?: PositioningProps;
}

export interface MenuProps {
  placement?: PositioningProps;
}

export interface PopoverProps {
  placement?: PositioningProps;
}
```

Example usage:

```tsx
type PlacementShorthand =
  | 'above'
  | 'above-start'
  | 'above-end'
  | 'below'
  | 'below-start'
  | 'below-end'
  | 'before'
  | 'before-top'
  | 'before-bottom'
  | 'after'
  | 'after-top'
  | 'after-bottom';

// Shorthand for `position` and `align` options
<PositionedComponent placement="above-start">

> <PositionedComponent placemant={{position: 'above', align: 'start'}}>

Reviewer: ```suggestion
<PositionedComponent placement={{position: 'above', align: 'start'}}>
```

<PositionedComponent placement="above">

> <PositionedComponent placemant={{position: 'above'}}>

Reviewer: ```suggestion
<PositionedComponent placement={{position: 'above'}}>
```

// Also allows more complex configuration
<PositionedComponent placement={{position: 'above', align: 'start', autoSize: true}}>
```

Slots that use popper should use a similar pattern if a top level `placement` option is not clear enough:

```tsx

> import { PositioningProps } from '@fluentui/react-positioning';
> 
> export interface ComponentWithPositionedSlotProps {
>   positionedSlot: React.HTMLAttributes<HTMLElement> & { placement?: PositioningProps };

Reviewer: ```suggestion
import { PlacementShorthand } from '@fluentui/react-positioning';

export interface ComponentWithPositionedSlotProps {
  positionedSlot: React.HTMLAttributes<HTMLElement> & { placement?: PlacementShorthand };
```

I guess this should also gave different type...

}
```

This way we can safely configure `usePopper` internally with a simple object spread that does not require knowing what
props are used by the component props:

```tsx

> // � Existing usage

Reviewer: ![image](https://user-images.githubusercontent.com/14183168/128847941-9211851d-39aa-4590-ba2d-dcadac7166af.png)

Is everything okay with this emoji? 🤨

// Need to be aware of user props and make sure they are added to the hook usage
usePopper({
  align: state.align,
  position: state.position,
  target: state.target,
  coverTarget: state.coverTarget,
  offset: state.offset,
});

// � Proposed usage
// Guaranteed to configure based on user props, and any component specific modifications after
usePopper({

>   ...placement,

Reviewer: Did not get how this will work if `placement` is a string. Can you please clarify?

# RFC:

@ling1726

## Summary

> In the v9 library, we have invested in [Popper.js](https://popper.js.org/) as the tool to managed positioned elements.

Reviewer 3: ```suggestion
In the v9 library, we have invested in [Popper.js](https://popper.js.org/) as the tool to manage positioned elements.
```

options that are required. This RFC proposes a way to configure popper uniformly in v9 components for consumers.

## Background

Configuring positioned props in both v0 and v8 are done inconsistently. Here are some examples

v8:

```tsx
<Callout
  {...props} // ICalloutProps interface
/>

<ContextualMenu
  calloutProps={calloutProps} // ICalloutProps interface
  // also can spread calloutProps
  calloutMaxHeight
  bounds
  directionalHintFixed
/>
```

v0:

```tsx
<Popup
  align
  position
  offset
/>

<Menu
  items={[
    { menu: popper: { align, position, offset }}
  ]}
/>
```

## Problem statement

There is a subset of `usePopper` options that are currently used by all converged components that need positioning.
These options are spread into the component's props:

```tsx
<Tooltip position="bottom" align="center" />
<Popover position="bottom" align="center" />
<Menu position="bottom" align="center" />
```

There are other popper options that are used in some components but not others, examples of these options are:

- `coverTarget` - allows the positioned element to cover the target

> There are also additional options that are not used by any converge component that might be used in the future like:

Reviewer 3: ```suggestion
There are also additional options that are not used by any converged component that might be used in the future like:
```

- `flipBoundary` - Determines what bounds to use to flip the positioning

Issues and bugs can be triggered because certain use cases require more precise configuration of popper. The most likely
result would be that a popper configuration is exposed for one component but not others.

## Detailed Design or Proposal

### Equal popper configuration

Each converged component that uses popper should allow an equal level of configuration. This ensures that our positioned
components can behave consistently across the library. Any fixes that require exposing another popper configuration will
be available to **all** components that use popper.

This proposal does not mean that all popper options must be public, but that if one component exposes a popper configuration,
then all components should do so equally.

We can achieve this with an agreement that the current `PositioningProps` would be supported by all v9 components, while
internal popper options can still belong to `PopperOptions` type in the `react-positioning` package.

### Shorthand `placement` configuration

The next step to this proposal is to make sure that all components which use a positioned element should allow users

> to configure popper through a `placement` prop that is behaves like a shorthand. An extra helper will be required to resolve

Reviewer 3: ```suggestion
to configure popper through a `placement` prop that behaves like a shorthand. An extra helper will be required to resolve
```

shorthand to full popper configuration internally, but the external API for `placement` is very clean for users.

The usage examples below will detail the usage of this shorthand.

```tsx
// @fluentui/react-positioning
export type PositioningShorthandValue =
  | 'above'
  | 'above-start'
  | 'above-end'
  | 'below'
  | 'below-start'
  | 'below-end'
  | 'before'
  | 'before-top'
  | 'before-bottom'
  | 'after'
  | 'after-top'
  | 'after-bottom';

> export type PositioningShorthand = PositioningProps | PositioningShorthandValue;

Reviewer 2: Should we consider renaming these types to align with the `placement` prop name?
`PlacementProps`
`PalcementShorthandValue`
`PlacementShorthand`

Author: well my original suggestion was `positioning` :)

> export function resolvePopperOptions(shorthand: PositioningShorthand, config: PositioningProps) {
>   if (typeof shorthand === 'object') {
>     return {
>       ...shorthand,
>       ...config,
>     };
>   }
> 
>   if (typeof shorthand === 'string') {
>     return {
>       ...parseStringShorthand(shorthand),
>       ...config,
>     };
>   }
> 
>   return config;
> }
> 
> export function parseStringShorthand(shorthand) {
>   return {
>     position,
>     align,
>   };
> }

Reviewer 2: It'd probably be better to let the components do the `{ ...shorthand, ...config }` spread themselves, rather than baking it into `resolvePopperOptions`.
* It doesn't support default values (`config` overrides the `placement` prop).
* Some of the values could be calculated based on the value of the placement prop. For example, Tooltip's `offset` value is `placement.offset + arrowHeight` (add a fixed amount to whatever the user passed in).
* Spreading an object is pretty natural and easy for someone to write if that's what they need:

I'd propose to delete `resolvePopperOptions` and _only_ have the `parseStringShorthand` function (maybe call it `resolvePositioningShorthand`?) to convert from shorthand to object.

Example usage might be:
```ts
const popperOptions = { offset: 4, ...resolvePositioningShorthand(state.placement) };
popperOptions.offset += arrowHeight;
const popper = usePopper(popperOptions);
```

--

```suggestion
export function resolvePositioningShorthand(placement: PositioningShorthand): PositioningShorthand;
export function resolvePositioningShorthand(placement: PositioningShorthand | undefined) : PositioningShorthand | undefined {
  if (typeof placement === 'string) {
    // ... parse the string, or lookup in a table ...
    return {
      position,
      align,
    };
  }
  
  return placement;
}
```

Author: I don't mind, parsing shorthand is always gonna be awkward, so it will have to be handled smwhere. More importantly the actual API for consumers will be greatly simplified.

I'm fine with going along with this proposal 👍

Author: In any case `usePopper` will continue to only accept a single object, and it **should not** accept shorthand values

Author: Also reverted to `positioning` it just fits well with everything currently there too

Reviewer 3: Highly approve `positioning` 👍
